### PR TITLE
feat(nnx): add preferred_element_type to attention module

### PR DESCRIPTION
**Pull request by aarushisingh04**
## What does this PR do?

adds `preferred_element_type` parameter to `dot_product_attention_weights`, `dot_product_attention` and `MultiHeadAttention` in `flax/nnx/nn/attention.py`.

split from [#5179](https://github.com/google/flax/pull/5179) for easier review.

**changes:**
- `dot_product_attention_weights`: forwards `preferred_element_type` to `jnp.einsum`
- `dot_product_attention`: forwards to both `jnp.einsum` calls and conditionally applies `.astype()` only when `preferred_element_type is not None`
- `MultiHeadAttention.__init__`: accepts `preferred_element_type` and passes it to all `LinearGeneral` layers (q/k/v/out)
- `MultiHeadAttention.__call__`: forwards `preferred_element_type` to `self.attention_fn`

**tests:**
- adds `test_preferred_element_type` to `attention_test.py`

## Checklist
- [ ] This PR fixes a minor issue (e.g.: typo or small bug) or improves the docs (you can dismiss the other checks if that's the case).
- [ ] This change is discussed in a Github issue/[discussion](https://github.com/google/flax/discussions) (please add a link).
- [x] The documentation and docstrings adhere to the [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [x] This change includes necessary high-coverage tests. (No quality testing = no merge!)

### CODE REVIEW

**Review by samanklesaria — APPROVED** [review]

**Review by IvyZX — COMMENTED** [review]
We already have `precision` to control the op precision and `dtype` to control the param precision. What's a use case for this additional API?

**Comment by samanklesaria** [comment]
> We already have `precision` to control the op precision and `dtype` to control the param precision. What's a use case for this additional API?

Good point. It seems like these parameters cover the use case for a `preferred_element_type` parameter. @aarushisingh04 does that make sense to you?

**Comment by aarushisingh04** [comment]
@IvyZX @samanklesaria 

thank you for reviewing, i'll address the mentioned points below-

i believe `precision` and `dtype` don't cover this use case. `precision` controls how carefully the hardware computes (e.g. `HIGHEST` uses more precise accumulators) but does not change the output `dtype`, as with bfloat16 inputs and `precision=Precision.HIGHEST`, you still get bfloat16 attention weights back. 

`dtype` promotes the inputs before the einsum but similarly doesn't let you accumulate in a higher dtype. so, `preferred_element_type` is the only way to get float32 attention logits when inputs are bfloat16, without manually inserting `.astype()` calls or casting the entire Q/K/V tensors to float32. this is the same design choice `FlashAttention` makes for numerical stability.

>just as a note, the motivation behind this PR was that `preferred_element_type` was already added to `nnx.Linear*`, `nnx.Conv*`, and `nnx.Einsum` in [PR #4920](https://github.com/google/flax/pull/4920), included in the Flax 0.12.0 release. this PR meant to extend the same pattern to `MultiHeadAttention` which was not covered by #4920.

if the pr is green-lighted after this discussion i will fix the docstrings (which currently incorrectly describe it as "numerical precision")

- preferred_element_type: numerical precision of the computation, see `jax.lax.dot_general` for details.

to a more accurate description

- preferred_element_type: Optional parameter controls the data type output by the dot product. This argument is passed to ``dot_general`` function. See ``jax.lax.dot`` for details.


and strengthen the tests to cover the mixed-precision use case.
